### PR TITLE
Add admin checksum table command

### PR DESCRIPTION
**Pull request by huachaohuang**
Support a new command `ADMIN CHECKSUM TABLE tablename` in TiDB.
The result of this command is a 64bit checksum, calculated from all key-values in a table, including records and indices.
This command will be implemented like the `ANALYZE table` command and will be pushed down to TiKV.
The corresponding TiKV PR is here: https://github.com/pingcap/tikv/pull/2822.

The current implementation use this checksum algorithm:
1. For every key-value in a table, including records and indices, calculate a CRC64 from (key + value).
2. For every CRC64 from every key-value, accumulate the checksum by XOR them together.

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -4778,6 +4778,13 @@ AdminStmt:
 			HandleRanges: $6.([]ast.HandleRange),
 		}
 	}
+|	"ADMIN" "CHECKSUM" "TABLE" TableNameList
+  {
+    $$ = &ast.AdminStmt{
+      Tp:	ast.AdminChecksumTable,
+      Tables: $4.([]*ast.TableName),
+    }
+  }
```
Comment (zimulala) [inline parser/parser.y]: Please align.

File: distsql/distsql.go [hunk 2]
```diff
@@ -344,6 +344,22 @@ func Analyze(ctx context.Context, client kv.Client, kvReq *kv.Request) (SelectRe
 	return result, nil
 }
 
+// Checksum do a checksum request.
```
Comment (zimulala) [inline distsql/distsql.go]: s/do/does

File: executor/checksum.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+func (c *checksumContext) buildTableRequest(ctx sessionctx.Context) ([]*kv.Request, error) {
+	checksum := &tipb.ChecksumRequest{
+		StartTs:   c.StartTs,
+		ScanOn:    tipb.ChecksumScanOn_Table,
+		Algorithm: tipb.ChecksumAlgorithm_Crc64_Xor,
+	}
+
+	var ranges []*ranger.NewRange
+	if c.TableInfo.PKIsHandle {
```
Comment (lamxTyler) [inline executor/checksum.go]: If the scan order does not matter for checksum, you can just use `FullIntNewRange(false)`.

File: executor/checksum.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+	var results []*checksumResult
+	for i := 0; i < len(tasks); i++ {
+		result := <-resultCh
+		if result.Error != nil {
+			err = result.Error
+			log.Error(errors.ErrorStack(err))
+			continue
+		}
+		results = append(results, result)
```
Comment (lamxTyler) [inline executor/checksum.go]: Why not handle the result once we receive one?

File: executor/checksum.go (hunk 3 above)
Comment (huachaohuang) [reply executor/checksum.go]: In my understanding, it still needs to handle the unsigned ranges even if it doesn't care about the order.

Comment (lamxTyler) [reply executor/checksum.go]: PTAL @winoros

**Comment by huachaohuang** [comment]
@zimulala @lamxTyler @winoros @winkyao PTAL, thanks

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -244,6 +244,9 @@ type SessionVars struct {
 	// BuildStatsConcurrencyVar is used to control statistics building concurrency.
 	BuildStatsConcurrencyVar int
 
+	// ChecksumTableConcurrencyVar is used to control checksum table concurrency.
+	ChecksumTableConcurrencyVar int
```
Comment (winkyao) [inline sessionctx/variable/session.go]: this variable is never used.

**Pull request merged**